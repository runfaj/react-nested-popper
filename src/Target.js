import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

export default class Target extends React.Component {
  static className = 'react-nested-popper_Target';


  /*  tracks ref so popperjs can be bound to it. Also provide to user if they want it.
      We wouldn't even need this component, but we need to guarantee a ref to the target,
      so we enforce this component.
  */
  setRef(el) {
    if (this.props.innerRef) {
      this.props.innerRef(el);
    }
    if (this.props._targetRef) {
      this.props._targetRef(el);
    }
  }

  render() {
    const rest = _.omit(this.props, _.keys(Target.propTypes));

    return (
      <div
        className={this.props.className}
        ref={el => this.setRef(el)}
        onClick={this.props._onClick}
        {...rest}
      >
        {this.props.children}
      </div>
    );
  }
}

// see docs for definitions of each external prop here
Target.propTypes = {
  // external props
  className: PropTypes.string,
  innerRef: PropTypes.func,

  // internal props - these can't be required unless you want it to show to end user
  _onClick: PropTypes.func,
  _targetRef: PropTypes.func,
};
Target.defaultProps = {
  className: '',
  innerRef: (el) => {},

  _onClick: (e) => {},
  _targetRef: (el) => {},
};
