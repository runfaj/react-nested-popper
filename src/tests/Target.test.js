import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Target from '../Target';

describe('Target component', () => {
  it('renders with nested text', () => {
    const component = renderer.create(
      <Target>react-nested-popper</Target>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with multiple children', () => {
    const component = renderer.create(
      <Target>
        <div>react-nested-popper</div>
        <div>react-nested-popper</div>
      </Target>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('calls both ref functions', () => {
    const setRef = jest.fn();
    const wrapper = mount(
      <Target
        innerRef={setRef}
        _targetRef={setRef}
      />
    );
    expect(setRef).toHaveBeenCalledTimes(2);
  });
});
