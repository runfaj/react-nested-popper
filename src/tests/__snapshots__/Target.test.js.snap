// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Target component renders with multiple children 1`] = `
<div
  className=""
  onClick={[Function]}
>
  <div>
    react-nested-popper
  </div>
  <div>
    react-nested-popper
  </div>
</div>
`;

exports[`Target component renders with nested text 1`] = `
<div
  className=""
  onClick={[Function]}
>
  react-nested-popper
</div>
`;
